sysdig/resource_sysdig_secure_cloud_auth_account.go

```diff
@@ -343,19 +343,26 @@ func constructAccountComponents(accountComponents []*cloudauth.AccountComponent,
 					}
 				case "service_principal_metadata":
 					// TODO: Make it more generic than just for GCP
-					service_principal_private_key := getServicePrincipalKeyObject(value.(string))
-					component.Metadata = &cloudauth.AccountComponent_ServicePrincipalMetadata{
-						ServicePrincipalMetadata: &cloudauth.ServicePrincipalMetadata{
-							Provider: &cloudauth.ServicePrincipalMetadata_Gcp{
-								Gcp: &cloudauth.ServicePrincipalMetadata_GCP{
-									Key: &cloudauth.ServicePrincipalMetadata_GCP_Key{
-										ProjectId:    data.Get("cloud_provider_id").(string),
-										PrivateKeyId: service_principal_private_key["private_key_id"],
-										PrivateKey:   service_principal_private_key["private_key"],
+					servicePrincipalMetadata := parseMetadataJson(value.(string))
```

Comment:
maybe you can refactor this and take provider as input for a new function?

Reply:
So, the way I have written this `parseMetadataJson` function is - its a common generic function which parses any metadata jsonified string and returns a map. For instance, you can even pass `trusted_role_metadata` to it and it will return a map of key:value pairs.

If I pass in the `provider` as well, then I'll also need to know which metadata I'm parsing for. This is because the metadata are different for different providers. We will likely end up having the same checks for parsing as this switch case, unless ofcourse we have different parsing functions for different metadata types.
Please let me know if I misunderstood the comment.

Reply:
I feel we will end up in if/else maze in this switch case. For now it's ok but maybe we want to revisit this.

Reply:
Yeah, I agree. I was equally concerned what if we end up having the if/else/switch maze in the new function too. Let me rethink through this and take it up later if we need to refactor.